feat: [#114] 경기 종료시 route 정보 저장

Damagucci-Juice opened the pull request:
## 🐲 관련 이슈

#114, 

## 🏃‍♂️ 구현/변경 사항
- [x] 코어로케이션 권한 획득 추가
- [x] 워치에서 현재 위치 읽어옴
- [x] 읽어 온 데이터가 저장됨

## 📷 스크린샷
![image](https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/50472122/c90af822-c851-43a2-9c7c-5f214fd971d0)

![Pasted Graphic](https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/50472122/2beff64e-a71c-4617-9b24-d302f8f424de)

## 🙏To Reviewer
@yunwkgus 코드 변경 사항이 `WorkoutManager`에서 생깁니다. 
리뷰시에 유의하면서 리뷰해주시면 감사하겠습니다. 

### 세션 저장 과정에서 흐름에 대한 무지
세션이 어떻게 멈추고 Builder에서 어떻게 저장하는지는 알겠는데 다음의 흐름을 모르겠습니다. 

1. `WorkoutManager.endWorkout()`이 먼저 불리는지, `WorkoutManager.workoutSession(_ workoutSession: HKWorkoutSession, didChangeTo toState: HKWorkoutSessionState,
                        from fromState: HKWorkoutSessionState, date: Date) `이 먼저 불리는지 순서 문제
2. `WorkoutManager.showingSummaryView`가 호출 되는 시점에 따라서 workoutSession에 영향이 가는지? 

이 부분이 해결이 안되서 저장이 안됩니다. 

---
# 11월 7일 추가 PR분 

## 작업내용
- [x] 읽어 온 데이터가 저장됨
- [x] `WorkoutManager.startWorkout()`이 두 번 호출되는 버그 수정

## 에러 발생의 이유
- `WorkoutManager.session`이 변화함에 따라서 `WorkoutManager.workoutSession`의 델리게이트 메서드가 작동을 해야하는데 어떠한 이유로 작동을 하지 않음
- 게임을 시작하든, 게임을 종료하든 session.running으로 동작함, (session.pause는 제대로 동작)
- 예상 동작은 "경기 종료" 버튼을 누르면 session.end 로 넘어갈 것을 기대했는데 이는 문제로 작용함
- 디버그 프린트를 찍으면서 디버그를 한 결과, 경기 종료 버튼을 누르면 WorkManager.startWorkout() 메서드가 중복해서 동작하는 것을 확인

## 버그 픽스 
1. `PrecountView`: 게임 시작 버튼을 누르면 나타나는 뷰인데, `startWorkout()`메서드를 `onDisapper()` 수정자에서 선언하고 있었음
2. 위 뷰를 `StartView`의 Body에서 if-else 문으로 선언하고 있는데, 문제는 경기 종료하면 StartView가 나타나 배경에 깔리고 그 위로 경기의 요약된 데이터를 덮어써서 보여주는 형식으로 보여주기 때문에 발생함. **다시 말해, 경기 종료시 `StartView-PrecountView`가 호출되는 것**
3. 결과적으로 1번에 언급한 `PrecountView`의 `onDisappear`가 호출되서 경기 종료시에도 `startWorkout()` 메서드가 호출됨
4. 따라서 새로운 세션이 생성되고 각종 빌더와 세션의 델리게이트가 새롭게 선언되어서 종료 메서드가 도달하지 않는 거였음
5. PrecountView가 사라질 때 호출하는 것이 아닌, count 의 값이 1이 될 때 startWorkout을 호출하도록 변경하였음

- 애플워치에서 종료 버튼을 누르는 화면
![일시정지, 종료](https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/50472122/337b7cf7-0166-40f7-ba73-8386a281624f)

- 경기 종료 후 화면 전환시 밑바탕에 깔리는 `StartView`, 0~1초에서 넘어가는 화면에서 밑에 깔리는 파란원이 `StartView`

https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/50472122/25d26b2f-9634-4abb-84ed-ee24d1e94817


- `startWorkout()`이 두 번 불리는 에러
<img width="1108" alt="start 두번 불리는 에러" src="https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/50472122/9457938a-1d96-4e4b-8c85-af18101d2960">

- `StartView`에서 if-else로 호출하는 `PrecountView`, 이 때문에 바로 disappear가 호출됨
<img width="761" alt="문제의 원인" src="https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/50472122/4bdca13a-ebb3-43fa-9853-249f64951a9c">

- `PrecountView`에서 사용하던 수정자를 변경
<img width="536" alt="이거를 이용해서 해결함" src="https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/50472122/86e95c24-05c5-4f5f-90ec-b7561bd09266">

- 해결되서 프린트문이 제대로 찍히는 모습
<img width="1259" alt="버그 해결해서 루트 저장되는 모습" src="https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/50472122/2650a1b4-1319-4c60-80b4-e6f5eb1c7cce">

Damagucci-Juice committed:
feat: [#114] 워치에서 위치 권한 획득 설정

Damagucci-Juice committed:
feat: [#114] 위치 정보 업데이트 되는 것 확인

Damagucci-Juice committed:
fix: [#114] 강제 언래핑 제거

Damagucci-Juice committed:
chore: [#114] 불필요한 프린트문 제거

Damagucci-Juice committed:
chore: [#114] 사용하지 않는 코드 제거

Damagucci-Juice committed:
chore: [#114] 필터된 로케이션의 수평 정확성 조정 필요 주석 추가

Damagucci-Juice committed:
refactor: [#114] 메타데이터 길게 늘어진것 변수로 빼냄

Damagucci-Juice committed:
refactor: [#114] 빌더의 작업 종료 코드를 HKWorkoutSessionDelegate 의 코드를 정의한 곳으로 이동

Damagucci-Juice committed:
feat: [#114] 세션이 End로 변하면 불리는 함수에서 변화를 보기 위해서 NSLog 선언

Damagucci-Juice committed:
refactor: [#114] Print문을 NSLog로 대체

yunwkgus commented:
<img width="242" alt="image" src="https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/52576276/715f12c8-09a2-400a-afd9-74b28e7ded34">
end workout 함수가 먼저 호출되어요!
실제로 종료 버튼을 누르면 바로 호출되는 함수라서 더 그런 것 같아요 -

yunwkgus commented:
<img width="289" alt="image" src="https://github.com/DeveloperAcademy-POSTECH/MacC-Team10-Guryongpo/assets/52576276/7073f9c8-eee1-44df-bd80-2163f045c42a">

showing summary 의 경우에는 @State bool 값이고 sheet 를 올려주는 값이라 workout session 과 영향은 크게 없을 것 같아요 ?? -마찬가지로 end workout 함수에서 변경되는 값이라 종료 버튼을 누르면 변경되는 값입니다.
오히려 session?.end() 이 함수가 workoutsession의 상태를 관리해주는 함수인데 이 함수의 영향으로 session의 상태가 .end 로 변경되고 -> workoutSession builder 에서 세션 상태 변경을 받아 WorkoutManager.workoutSession 함수가 호출될 것이라 생각합니다

Damagucci-Juice committed:
feat: [#114] Save New Route

Damagucci-Juice committed:
refactor: [#114] SwiftLint 규칙 준수 위해 3항 연산 줄 바꿈

Damagucci-Juice committed:
chore: [#114] Workout의 세션이 변화하는 곳을 나타내는 주석 추가

Damagucci-Juice committed:
feat: [#114] 위치 권한 업데이트 코드가 두번 불리는 부분 수정

Damagucci-Juice committed:
chore: [#114] Location을 확인하는 주석 추가

- 추후에 이 코드가 필요없다고하면 제거 할 수도 있음
- HealthKit에서 기본적으로 위치 정보를 제공하는 것 같은데 확인차 CoreLocation을 담은 거라서 삭제 가능성 있음

Damagucci-Juice committed:
fix: [#114] StartWorkout() 메서드가 두 번 호출되는 버그를 수정

Damagucci-Juice committed:
setting: [#114] Apple Watch 쪽 info.plist에 위치 정보 획득 권한 명시

Damagucci-Juice commented:
> <img alt="image" width="289" src="https://user-images.githubusercontent.com/52576276/280454148-7073f9c8-eee1-44df-bd80-2163f045c42a.png">
> showing summary 의 경우에는 @State bool 값이고 sheet 를 올려주는 값이라 workout session 과 영향은 크게 없을 것 같아요 ?? -마찬가지로 end workout 함수에서 변경되는 값이라 종료 버튼을 누르면 변경되는 값입니다. 오히려 session?.end() 이 함수가 workoutsession의 상태를 관리해주는 함수인데 이 함수의 영향으로 session의 상태가 .end 로 변경되고 -> workoutSession builder 에서 세션 상태 변경을 받아 WorkoutManager.workoutSession 함수가 호출될 것이라 생각합니다

저도 이 흐름이 일반적이라 느껴서 궁금했는데, 11월 7일 추가분 PR에 명시했듯이 버그가 숨어있었네요! 감사해요!

yunwkgus commented:
대단합니다 -! 멋지네요 -
